### ISSUE TITLE
Segmentation faults in DataLoader (in latest torch version).

### ISSUE THREAD

**Issue by magical-inference** [pytorch/pytorch#91245]
### 🐛 Describe the bug

I'm experiencing segmentation faults in the dataloader.

# To Reproduce

This happens when `num_workers=16`, potentially also with other integers > 0. 

The Dataloader doesn't segfault when num_workers isn't set.

# Expected behavior

No segfaults

# Relevant snippet from code
```python
dataloader = torch.utils.data.DataLoader(train_data, shuffle=True,
                                         batch_size=batch_size,
                                         num_workers=16)
```

# Trace

```
ERROR: Unexpected segmentation fault encountered in worker.
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Cell In[20], line 38
     36 errG = criterion(output, labels)
     37 errG.backward()
---> 38 D_G_z2 = output.mean().item()
     39 optimizerG.step()
     41 # Save Losses for plotting later

File /lab/mlenv/lib/python3.10/site-packages/torch/utils/data/_utils/signal_handling.py:66, in _set_SIGCHLD_handler.<locals>.handler(signum, frame)
     63 def handler(signum, frame):
     64     # This following call uses `waitid` with WNOHANG from C side. Therefore,
     65     # Python can still get and update the process status successfully.
---> 66     _error_if_any_worker_fails()
     67     if previous_handler is not None:
     68         assert callable(previous_handler)

RuntimeError: DataLoader worker (pid 139803) is killed by signal: Segmentation fault. 
```

I've skimmed through earlier issues but those were closed and apparently fixed.

kind regards
Max

### Versions

PyTorch version: 1.13.1+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.1 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.10.6 (main, Nov 14 2022, 16:10:14) [GCC 11.3.0] (64-bit runtime)
Python platform: Linux-5.15.0-56-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4090
Nvidia driver version: 525.60.11
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.24.0
[pip3] torch==1.13.1
[pip3] torchvision==0.14.1
[conda] Could not collect

**Comment by drisspg** [comment]
Can you provide a minimal script that reproduces the error?

**Comment by magical-inference** [comment]
Here's a full md-dump of the notebook, including outputs without images.
[GAN_dev_1.md](https://github.com/pytorch/pytorch/files/10295852/GAN_dev_1.md), sorry for not cleaning up, I don't have the time at the moment.

**Comment by ejguan** [comment]
Are you running it in notebook?

I highly suspect it's caused by running out of cuda shared memory.

**Comment by magical-inference** [comment]
I'm running it in notebook, yes.

The process never consumed more than ~30% of my gpus memory (according to nvidia-smi).
Am I missing something here? Like a config-param to make all gpu-memory available for the process?

Btw, while debugging I also got segfaults while setting torch.device to "cpu" (while keeping num_workers > 1).

**Comment by ejguan** [comment]
Can you check how much shared memory used? The more workers the more shared memory is needed.

**Comment by magical-inference** [comment]
I have no access to the machine for about two weeks, starting today.

If shared memory is equivalent to gpu memory in this context (I'm rather new to CUDA and don't fully understand the term after briefly looking into it): up to ~8Gb of 22 available were used.

**Comment by ejguan** [comment]
Nope. Shared memory is used to pass Tensor from worker process to the main process. 

To check shared memory usage, you can do `ipcs -pm`
You can find the limit of your shared memory using `cat /proc/sys/kernel/shmmax`.
And, you can resize it using `echo <bytes> >  /proc/sys/kernel/shmmax`.

**Comment by magical-inference** [comment]
Thank you for the explanation, I get shared memory now.
I'll look into it as soon as the machine is available again.

In case this helps with the investigation: all this happened on a fresh Ubuntu 22.04 server install using the latest nvidia-drivers (`525`), so if installing the nvidia-drivers (clean) sets a reasonable limit for shared memory, that's not the culprit.

**Comment by Vasud-ha** [comment]
I tested this code on ubuntu==20.04 with Nvidia drivers (510) for 150 epochs and didn't face any errors.

**Comment by Vasud-ha** [comment]
@magical-inference Can you please confirm if you face this issue on the CPU device or the Cuda device, as the code snippet you shared is meant for Cuda tensors?
